File: core/core-web/src/main/java/ch/difty/scipamato/core/web/common/DefinitionEditPage.java
package ch.difty.scipamato.core.web.common;

import lombok.extern.slf4j.Slf4j;
import org.apache.wicket.PageReference;
import org.apache.wicket.markup.html.basic.Label;
import org.apache.wicket.markup.html.form.Form;
import org.apache.wicket.model.CompoundPropertyModel;
import org.apache.wicket.model.IModel;
import org.apache.wicket.model.StringResourceModel;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.springframework.dao.DuplicateKeyException;

import ch.difty.scipamato.common.entity.DefinitionEntity;
import ch.difty.scipamato.core.persistence.OptimisticLockingException;

@Slf4j
@SuppressWarnings({ "SameParameterValue", "WicketForgeJavaIdInspection", "SpellCheckingInspection" })
public abstract class DefinitionEditPage<E extends DefinitionEntity<?, ?>> extends BasePage<E> {

    private final PageReference callingPageRef;

    private Form<E> form;

    protected DefinitionEditPage(@Nullable final IModel<E> model, @Nullable final PageReference callingPageRef) {
        super(model);
        this.callingPageRef = callingPageRef;
    }

    @Nullable
    protected PageReference getCallingPageRef() {
        return callingPageRef;
    }

    @NotNull
    protected Form<E> getForm() {
        return form;
    }

    @Override
    protected final void onInitialize() {
        super.onInitialize();
        queueForm("form");
    }

    private void queueForm(@NotNull final String id) {
        queue(form = newForm(id));

        queue(new Label("headerLabel", new StringResourceModel("header" + LABEL_RESOURCE_TAG, this, null)));
        queue(newDefinitionHeaderPanel("headerPanel"));

        queue(new Label("translationsLabel", new StringResourceModel("translations" + LABEL_RESOURCE_TAG, this, null)));
        queue(newDefinitionTranslationPanel("translationsPanel"));
    }

    private Form<E> newForm(@NotNull final String id) {
        return new Form<>(id, new CompoundPropertyModel<>(getModel())) {
            @Override
            protected void onSubmit() {
                super.onSubmit();
                try {
                    E persisted = persistModel();
                    if (persisted != null) {
                        setModelObject(persisted);
                        info(new StringResourceModel("save.successful.hint", this, null)
                            .setParameters(getModelObject().getNullSafeId(), getModelObject().getTranslationsAsString())
                            .getString());
                    } else {
                        handleRepoException();
                    }
                } catch (OptimisticLockingException ole) {
                    handleOptimisticLockingException(ole);
                } catch (DuplicateKeyException dke) {
                    handleDuplicateKeyException(dke);
                } catch (Exception ex) {
                    handleOtherException(ex);
                }
            }

            private void handleRepoException() {
                error(new StringResourceModel("save.unsuccessful.hint", this, null)
                    .setParameters(getModelObject().getNullSafeId(), "")
                    .getString());
            }

            private void handleOptimisticLockingException(final OptimisticLockingException ole) {
                final String msg = new StringResourceModel("save.optimisticlockexception.hint", this, null)
                    .setParameters(ole.getTableName(), getModelObject().getNullSafeId())
                    .getString();
                log.error(msg);
                error(msg);
            }

            private void handleOtherException(final Exception oe) {
                final String msg = new StringResourceModel("save.error.hint", this, null)
                    .setParameters(getModelObject().getNullSafeId(), oe.getMessage())
                    .getString();
                log.error(msg);
                error(msg);
            }
        };
    }

    @Nullable
    protected abstract E persistModel();

    @NotNull
    protected abstract DefinitionEditHeaderPanel<E, ?, ?> newDefinitionHeaderPanel(@NotNull final String id);

    @NotNull
    protected abstract DefinitionEditTranslationPanel<E, ?> newDefinitionTranslationPanel(@NotNull final String id);

    protected abstract void handleDuplicateKeyException(@NotNull final DuplicateKeyException dke);
}

File: core/core-web/src/main/java/ch/difty/scipamato/core/web/common/DefinitionEditPage.kt
package ch.difty.scipamato.core.web.common

import ch.difty.scipamato.common.entity.DefinitionEntity
import ch.difty.scipamato.common.entity.DefinitionTranslation
import ch.difty.scipamato.common.logger
import ch.difty.scipamato.common.web.LABEL_RESOURCE_TAG
import ch.difty.scipamato.core.persistence.OptimisticLockingException
import org.apache.wicket.PageReference
import org.apache.wicket.markup.html.basic.Label
import org.apache.wicket.markup.html.form.Form
import org.apache.wicket.model.CompoundPropertyModel
import org.apache.wicket.model.IModel
import org.apache.wicket.model.StringResourceModel
import org.springframework.dao.DuplicateKeyException

private val log = logger()

@Suppress("SameParameterValue")
abstract class DefinitionEditPage<E : DefinitionEntity<ID, T>, T : DefinitionTranslation, ID>(
    model: IModel<E>?,
    protected val callingPageRef: PageReference?,
) : BasePage<E>(model) {

    private lateinit var form: Form<E>

    protected fun getForm(): Form<E> {
        return form
    }

    override fun onInitialize() {
        super.onInitialize()
        queueForm("form")
    }

    private fun queueForm(id: String) {
        queue(newForm(id).also { form = it })
        queue(Label("headerLabel", StringResourceModel("header$LABEL_RESOURCE_TAG", this, null)))
        queue(newDefinitionHeaderPanel("headerPanel"))
        queue(Label("translationsLabel", StringResourceModel("translations$LABEL_RESOURCE_TAG", this, null)))
        queue(newDefinitionTranslationPanel("translationsPanel"))
    }

    @Suppress("TooGenericExceptionCaught")
    private fun newForm(id: String): Form<E> = object : Form<E>(id, CompoundPropertyModel(model)) {
        override fun onSubmit() {
            super.onSubmit()
            try {
                val persisted = persistModel()
                if (persisted != null) {
                    modelObject = persisted
                    info(StringResourceModel("save.successful.hint", this, null)
                        .setParameters(modelObject!!.nullSafeId, modelObject!!.translationsAsString).string)
                } else {
                    handleRepoException()
                }
            } catch (ole: OptimisticLockingException) {
                handleOptimisticLockingException(ole)
            } catch (dke: DuplicateKeyException) {
                handleDuplicateKeyException(dke)
            } catch (ex: Exception) {
                handleOtherException(ex)
            }
        }

        private fun handleRepoException() {
            error(StringResourceModel("save.unsuccessful.hint", this, null)
                .setParameters(modelObject!!.nullSafeId, "").string)
        }

        private fun handleOptimisticLockingException(ole: OptimisticLockingException) {
            val msg = StringResourceModel("save.optimisticlockexception.hint", this, null)
                .setParameters(ole.tableName, modelObject!!.nullSafeId).string
            log.error(msg)
            error(msg)
        }

        private fun handleOtherException(oe: Exception) {
            val msg = StringResourceModel("save.error.hint", this, null)
                .setParameters(modelObject!!.nullSafeId, oe.message)
                .string
            log.error(msg)
            error(msg)
        }
    }

    protected abstract fun persistModel(): E?
    protected abstract fun newDefinitionHeaderPanel(id: String): DefinitionEditHeaderPanel<E, T, ID>
    protected abstract fun newDefinitionTranslationPanel(id: String): DefinitionEditTranslationPanel<E, T>
    protected abstract fun handleDuplicateKeyException(dke: DuplicateKeyException)

    companion object {
        private const val serialVersionUID: Long = 1
    }
}
